File: cpp/src/detection-formats.cpp
#include <detection-formats.h>

#include <detection.h>
#include <retract.h>
#include <pick.h>
#include <correlation.h>
#include <stationInfo.h>
#include <stationInfoRequest.h>

#include <string>
#include <vector>

namespace detectionformats {
////////////// functions //////////////

bool ValidateJSON(const std::string & jsonstring) {
    std::string errorstring = ValidateJSONErrors(jsonstring);

	if (errorstring.length() == 0) {
        return(true);
    } else {
        return(false);
    }
}

std::string ValidateJSONErrors(const std::string & jsonstring) {
    rapidjson::Document jsondocument;

	// parse the json into a document
	if (jsondocument.Parse(jsonstring.c_str()).HasParseError()) {
		return ("Parse Error");
	}

	int type = GetDetectionType(jsondocument);

	if (type == formattypes::unknown) {
		return("Unknown Format");
	}

    std::vector<std::string> errorlist;
    std::string errorstring = "";

	if (type == formattypes::picktype) {
        detectionformats::pick testpick(jsondocument);
        errorlist = testpick.geterrors();
	} else if (type == formattypes::correlationtype) {
        detectionformats::correlation testcorrelation(jsondocument);
        errorlist = testcorrelation.geterrors();
	} else if (type == formattypes::detectiontype) {
        detectionformats::detection testdetection(jsondocument);
        errorlist = testdetection.geterrors();
	} else if (type == formattypes::retracttype) {
        detectionformats::retract testretract(jsondocument);
        errorlist = testretract.geterrors();
	} else if (type == formattypes::stationinfotype) {
        detectionformats::stationInfo teststationinfo(jsondocument);
        errorlist = teststationinfo.geterrors();
	} else if (type == formattypes::stationinforequesttype) {
        detectionformats::stationInfoRequest testrequest(jsondocument);
        errorlist = testrequest.geterrors();
	}

    // no errors
    if (errorlist.size() == 0) {
		return ("");
	}

    for (int i = 0; i < errorlist.size(); i++) {
        errorstring += " " + errorlist[i];
    }

    return(errorstring);
}

}  // namespace detectionformats

File: java/src/main/java/gov/usgs/detectionformats/DetectionFormats.java
package gov.usgs.detectionformats;

import java.util.ArrayList;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

/**
 * a general class containing functions global to detectionformats.
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class DetectionFormats {

    /**
	 * Checks to see if the provided json string is valid
	 * 
	 * @param jsonString
	 *            the json formatted string to check.
	 * @return Returns a true if valid, false otherwise
	 */
    public static boolean validateJSON(String jsonString) {
        String errorString = validateJSONErrors(jsonString);
    
        // it's valid if there are no errors
        if (errorString.length() == 0) {
            return(true);
        } else {
            return(false);
        }
    }

    /**
	 * Gets any validation errors in the provided json string.
	 * @param jsonString
	 *            the json object to check.
	 * @return Returns a String containing all errors found
	 */
    public static String validateJSONErrors(String jsonString) {
        // use a parser to convert to a string
		JSONParser parser = new JSONParser();

		// parse it
		JSONObject newJSONObject;
		try {
			newJSONObject = (JSONObject) parser.parse(jsonString);
		} catch (ParseException e) { 
			return("Bad Format");
		}

        // check type
        FormatTypes type = Utility.getDetectionType(newJSONObject);
        if (type == FormatTypes.Unknown) {
            return("Unknown Format");
        }

        // get errors based on type
        ArrayList<String> errorList = new ArrayList<String>();
        String errorString = new String();
        if (type == FormatTypes.Pick) {
            Pick testPick = new Pick(newJSONObject);
            errorList = testPick.getErrors();
        } else if (type == FormatTypes.Correlation) {
            Correlation testCorrelation = new Correlation(newJSONObject);
            errorList = testCorrelation.getErrors();
        } else if (type == FormatTypes.Detection) {
            Detection testDetection = new Detection(newJSONObject);
            errorList = testDetection.getErrors();
        } else if (type == FormatTypes.Retract) {
            Retract testRetract = new Retract(newJSONObject);
            errorList = testRetract.getErrors();
        } else if (type == FormatTypes.StationInfo) {
            StationInfo testStationInfo = new StationInfo(newJSONObject);
            errorList = testStationInfo.getErrors();
        } else if (type == FormatTypes.StationInfoRequest) {
            StationInfoRequest testRequest = new StationInfoRequest(newJSONObject);
            errorList = testRequest.getErrors();
        }

        // no errors
        if (errorList.size() == 0) {
            return ("");
        }

        // combine the errors into a single string
        for (int i = 0; i < errorList.size(); i++) {
            errorString += " " + errorList.get(i);
        }

        return(errorString);
    }
}
